README.md
Stocator - Storage Connector for Apache Spark
==============================
Apache Spark can work with multiple data sources that include various object stores like IBM Cloud Object Storage, OpenStack Swift and more. To access an object store, Apache Spark uses Hadoop modules that contain connectors to the various object stores. 

Apache Spark needs only small set of object store functionalities. Specifically, Apache Spark requires object listing, objects creation, read objects, and getting data partitions. Hadoop connectors, however, must be compliant with the Hadoop ecosystem. This means they support many more operations, such as shell operations on directories, including move, copy, rename, etc. (these are not native object store operations). Moreover, Hadoop Map Reduce Client is designed to work with file systems and not object stores. The temp files and folders it uses for every write operation are renamed, copied, and deleted. This leads to dozens of useless requests targeted at the object store. It’s clear that Hadoop is designed to work with file systems and not object stores.

Stocator is implicitly designed for the object stores, it has very a different architecture from the existing Hadoop connector. It doesn’t depends on the Hadoop modules and interacts directly with object stores.

Stocator is a generic connector, that may contain various implementations for object stores. It shipped with OpenStack Swift and IBM Cloud Object Storage connectors. Stocator can be easily extended with more object store implementations.

## Major features
* Hadoop ecosystem compliant. Implements Hadoop FileSystem interface
* No need to change or recompile Apache Spark
* Stocator doesn’t create any temporary folders or files for write operations. Each Spark's task generates only one object in the object store. Preserves existing Hadoop fault tolerance model
* Object's name may contain "/" and thus simulate directory structures
* Containers / buckets are automatically created
* Supports speculate mode
* Stocator uses Apache httpcomponents.httpclient.version version 4.5.2 and up
 

## Stocator build procedure

* Checkout the master branch `https://github.com/SparkTC/stocator.git`
* Change directory to `stocator`
* Execute `mvn install`
* If you want to build a jar with all the dependencies, please execute 

		mvn clean package -Pall-in-one

## Usage with Apache Spark
Stocator can be used easily with Apache Spark. There are different ways to use Stocator

### Using spark-packages
Stocator deployed on spark-packages. This is the easiest form to integrate Stocator in Spark.
Just follow [stocator](https://spark-packages.org/package/SparkTC/stocator)

### Using Stocator without compiling Apache Spark

It is possible to execute Apache Spark with Stocator, without compiling Apache Spark. Just make sure that Stocator jars on the class path. Build Stocator with 

		mvn clean package -Pall-in-one

Directory `stocator/target` contains standalone jar `stocator-X.Y.Z-jar-with-dependencies.jar`.

The best is to extend Spark's class path to include Stocator jar. Edit `conf/spark-defaults.conf` and add Stocator to the class path. For example

	spark.driver.extraClassPath=/<PATH>/stocator/target/stocator-X-Y-Z-jar-with-dependencies.jar
	spark.executor.extraClassPath=/<PATH>/stocator/target/stocator-X-Y-Z-jar-with-dependencies.jar

	 
Another option is to run Apache Spark with 

	./bin/spark-shell --jars stocator-X.Y.Z-jar-with-dependencies.jar
However this is less adviced, as Spark will copy Stocator jar to the executors, which may consume time.

### Using Stocator with Apache Spark compilation

Less recommended, only for advanced users who want to recompile Spark. Both main `pom.xml` and `core/pom.xml` of Spark should be modified.
 add to the `<properties>` of the main pom.xml
	
	<stocator.version>X.Y.Z</stocator.version>

 add `stocator` dependency to the main pom.xml

     <dependency>
          <groupId>com.ibm.stocator</groupId>
          <artifactId>stocator</artifactId>
          <version>${stocator.version}</version>
          <scope>${hadoop.deps.scope}</scope>
      </dependency>

modify `core/pom.xml` to include `stocator`

    <dependency>
          <groupId>com.ibm.stocator</groupId>
          <artifactId>stocator</artifactId>
     </dependency>

	
Compile Apache Spark with Haddop support as described [here](https://spark.apache.org/docs/latest/building-spark.html)

## General requirements
Stocator verifies that 

	mapreduce.fileoutputcommitter.marksuccessfuljobs=true

If not modified, the default value of `mapreduce.fileoutputcommitter.marksuccessfuljobs` is `true`.

## Configuration keys
Stocator uses configuration keys that can be configured via spark's `core-site.xml` or provided in run time without using `core-site.xml`. To provide keys in run time use SparkContext variable with

	sc.hadoopConfiguration.set("KEY","VALUE")

For usage with `core-site.xml`, see the configuration template located under `conf/core-site.xml.template`.


## Stocator and IBM Cloud Object Storage (IBM COS)
Stocator allows to access IBM Cloud Object Storage via `cos://` schema. The general URI is the form

	cos://<bucket>.<service>/object(s)

where `bucket` is object storage bucket and `<service>` identifies configuration group entry.

### Using multiple service names
Each `<service>` may be any text, without special characters. Each service may use it's specific credentials and has different endpoint. By using multiple `<service>` allows to use different endpoints simultaneously. 

For example,  `service=myObjectStore`, then URI will be of the form

	cos://<bucket>.myObjectStore/object(s)

and configuration keys will have prefix `fs.cos.myObjectStore`.
If none provided, the default value is `service`, e.g.

	cos://<bucket>.service/object(s)

and configuration keys will have prefix `fs.cos.service`.


### Reference Stocator in the core-site.xml

Configure Stocator in `conf/core-site.xml`

	<property>
		<name>fs.stocator.scheme.list</name>
		<value>cos</value>
	</property>
	<property>
		<name>fs.cos.impl</name>
		<value>com.ibm.stocator.fs.ObjectStoreFileSystem</value>
	</property>
	<property>
		<name>fs.stocator.cos.impl</name>
		<value>com.ibm.stocator.fs.cos.COSAPIClient</value>
	</property>
	<property>
		<name>fs.stocator.cos.scheme</name>
		<value>cos</value>
	</property>

### Configuration keys
Stocator COS connector expose "fs.cos." keys prefix. For backward compatibility Stocator also supports "fs.s3d" and "fs.s3a" prefix, where "fs.cos" has the highest priority and will overwrite other keys, if present.

#### COS Connector configuration with IAM
To work with IAM and provide `api key` please switch to the relevant `ibm-sdk` branch depends on the Stocator version you need. For example for Stocator 1.0.24 release, switch to `1.0.24-ibm-sdk`, for Stocator master `1.0.25-SNAPSHOT`, switch to `1.0.25-SNAPSHOT-IBM-SDK` and so on.

You will need to build Stocator manually, for example using 1.0.24-ibm-sdk branch:

	git clone https://github.com/SparkTC/stocator
	cd stocator
	git fetch
	git checkout -b 1.0.24-ibm-sdk origin/1.0.24-ibm-sdk
	mvn clean install –DskipTests

You now need to include the `target/stocator-1.0.24-SNAPSHOT-IBM-SDK.jar` into class path of Spark. Follow section Using Stocator without compiling Apache Spark

##### Configure Stocator

The next step if to configure Stocator with your COS credentials. The COS credentials is of the form

	{
	  "apikey": "123",
	  "endpoints": "https://cos-service.bluemix.net/endpoints",
	  "iam_apikey_description": "Auto generated apikey during resource-key operation for Instance - abc",
	  "iam_apikey_name": "auto-generated-apikey-123",
	  "iam_role_crn": "role",
	  "iam_serviceid_crn": "identity-123::serviceid:ServiceId-XYZ",
	  "resource_instance_id": "abc"
	}

The following is the list of the Stocator configuration keys. `<service>` can be any value, for example `myCOS`

| Key | Info | Mandatory | value |
| --- | ------------ | ------------- | --------|
|fs.cos.`<service>`.iam.api.key | API key  | mandatory | value of `apiKey`
|fs.cos.`<service>`.iam.service.id  | Service ID | mandatory | Value of `iam_serviceid_crn`. In certain cases you need only value after `:serviceid:`
|fs.cos.`<service>`.endpoint | COS endpoint | mandatory | Open link from `endpoints` and choose relevant endpoint. This endpoint should go here
|fs.cos.`<service>`.iam.token.max.retry | IAM token retrieval max retry  | Optional |
|fs.cos.`<service>`.iam.token.refresh | IAM token refresh time offset in seconds  | Optional |


Example, configure `<service>` as `myCOS`:

	<property>
		<name>fs.cos.myCos.iam.api.key</name>
		<value>123</value>
	</property>
	<property>
		<name>fs.cos.myCos.endpoint</name>
		<value>http://s3-api.us-geo.objectstorage.softlayer.net</value>
	</property>
	<property>
		<name>fs.cos.myCos.iam.service.id</name>
		<value>ServiceId-XYZ</value>
	</property>

Now you can use URI

	cos://mybucket.myCos/myobject(s)


An optional, it is possible to provide existing token instead of using API key.
Instead of providing `fs.cos.myCos.iam.api.key`, Stocator supports `fs.cos.myCos.iam.token` that may contain value of the existing token. When token is expired, Stocator will throw `403` exception. It's the user responsibility to provide long activation token or re-create token outside of Stocator.

#### COS Connector configuration without IAM

The following is the list of the configuration keys. `<service>` can be any value, for example `myCOS`

| Key | Info | Mandatory |
| --- | ------------ | ------------- |
|fs.cos.`<service>`.access.key | Access key  | mandatory
|fs.cos.`<service>`.secret.key  | Secret key | mandatory
|fs.cos.`<service>`.endpoint | COS endpoint | mandatory
|fs.cos.`<service>`.v2.signer.type |  Signer type  | optional

Example, configure `<service>` as `myCOS`:

	<property>
		<name>fs.cos.myCos.access.key</name>
		<value>ACCESS KEY</value>
	</property>
	<property>
		<name>fs.cos.myCos.endpoint</name>
		<value>http://s3-api.us-geo.objectstorage.softlayer.net</value>
	</property>
	<property>
		<name>fs.cos.myCos.secret.key</name>
		<value>SECRET KEY</value>
	</property>
	<property>
		<name>fs.cos.myCos.v2.signer.type</name>
		<value>false</value>
	</property>

Now you can use URI

	cos://mybucket.myCos/myobject(s)
#### COS Connector optional configuration tuning


| Key | Default | Info |
| --- | ------- | ------ |
| fs.cos.socket.send.buffer | 8*1024 | socket send buffer to be used in the client |
| fs.cos.socket.recv.buffer | 8*1024 |socket send buffer to be used in the client |
| fs.cos.paging.maximum| 5000 | number of records to get while paging through a directory listing |
| fs.cos.threads.max | 10 | the maximum number of threads to allow in the pool used by TransferManager |
| fs.cos.threads.keepalivetime| 60 |the time an idle thread waits before terminating |
| fs.cos.signing-algorithm | | override signature algorithm used for signing requests |
| fs.cos.connection.maximum| 10000 | number of simultaneous connections to the object store |
| fs.cos.attempts.maximum| 20 | number of times we should retry errors |
| fs.cos.block.size| 128 | size of a block in MB |
| fs.cos.connection.timeout | 800000 | amount of time (in ms) until we give up on a connection to the object store |
| fs.cos.connection.establish.timeout| 50000 | amount of time (in ms) until we give up trying to establish a connection to the object store |
| fs.cos.client.execution.timeout| 500000 | amount of time (in ms) to allow a client to complete the execution of an API call |
| fs.cos.client.request.timeout| 500000 | amount of time to wait (in ms) for a request to complete before giving up and timing out |
|fs.cos.connection.ssl.enabled |true | Enables or disables SSL connections to COS |
|fs.cos.proxy.host| | Hostname of the (optional) proxy server for COS connections |
|fs.cos.proxy.port| |Proxy server port. If this property is not set but fs.cos.proxy.host is, port 80 or 443 is assumed (consistent with the value of fs.cos.connection.ssl.enabled) |
| fs.cos.proxy.username| |Username for authenticating with proxy server |
| fs.cos.proxy.password| |Password for authenticating with proxy server |
| fs.cos.proxy.domain| |Domain for authenticating with proxy server |
| fs.cos.user.agent.prefix| |User agent prefix |
| fs.cos.flat.list | true | In flat listing the result will include all objects under specific path prefix, for example bucket/a/b/data.txt, bucket/a/d.data. If listed bucket/a*, then result will include both objects. If flat list is set to flase, then it contains the same list behaviour as community s3a connector. |
| fs.stocator.cache.size | 2000 | The Guava cache size used by the COS connector |
| fs.cos.multipart.size | 104857600 | Size in bytes. Define multipart size |
| fs.cos.multipart.threshold | Max Integer | minimum size in bytes before we start a multipart uploads, default is max integer |
| fs.cos.fast.upload | false | enable or disable block upload |
| fs.stocator.glob.bracket.support | false | if true supports Hadoop string patterns of the form {ab,c{de, fh}}. Due to possible collision with object names, this mode prevents from create an object whose name contains {} |

## Stocator and Object Storage based on OpenStack Swift API

Stocator allows to access OpenStack Swift API based object stores via unique schema `swift2d://`. 

* Uses streaming for object uploads, without knowing object size. This is unique to Swift connector and removes the need to store object locally prior uploading it.
* Supports Swiftauth, Keystone V2, Keystone V3 Password Scope Authentication
* Supports any object store that expose Swift API and supports different authentication models
* Supports access to public containers.  For example

		sc.textFile("swift2d://dal05.objectstorage.softlayer.net/v1/AUTH_ID/CONT/data.csv")


### Configure Stocator in the core-site.xml

Add the dependence to Stocator in `conf/core-site.xml`

	<property>
		<name>fs.stocator.scheme.list</name>
		<value>swift2d</value>
	</property>

If Swift connector used concurrently with COS connector, then also configure 

	<property>
		<name>fs.stocator.scheme.list</name>
		<value>swift2d,cos</value>
	</property>

Configure the rest of the keys

	<property>
		<name>fs.swift2d.impl</name>
		<value>com.ibm.stocator.fs.ObjectStoreFileSystem</value>
	</property>
	<property>
		<name>fs.stocator.swift2d.impl</name>
		<value>com.ibm.stocator.fs.swift.SwiftAPIClient</value>
	</property>
	<property>
		<name>fs.stocator.swift2d.scheme</name>
		<value>swift2d</value>
	</property>

	
   	
#### Swift connector configuration
The following is the list of the configuration keys

| Key | Info | Default value |
| --- | ------------ | ------------- |
|fs.swift2d.service.SERVICE_NAME.auth.url | Mandatory |
|fs.swift2d.service.SERVICE_NAME.public | Optional. Values: true, false | false |
|fs.swift2d.service.SERVICE_NAME.tenant | Mandatory |
|fs.swift2d.service.SERVICE_NAME.password |  Mandatory |
|fs.swift2d.service.SERVICE_NAME.username | Mandatory |
|fs.swift2d.service.SERVICE_NAME.block.size | Block size in MB | 128MB |
|fs.swift2d.service.SERVICE_NAME.region | Mandatory for Keystone|
|fs.swift2d.service.SERVICE_NAME.auth.method | Optional. Values: keystone, swiftauth, keystoneV3| keystoneV3
|fs.swift2d.service.SERVICE_NAME.nonstreaming.upload | Optional. If set to true then any object upload will be stored locally in the temp file and uploaded on close method. Disable stocator streaming mode | false



#### Example of core-site.xml keys
##### Keystone V2

	<property>
       <name>fs.swift2d.service.SERVICE_NAME.auth.url</name>
    	<value>http://IP:PORT/v2.0/tokens</value>
	</property>
	<property>
       <name>fs.swift2d.service.SERVICE_NAME.public</name>
    	<value>true</value>
	</property>
	<property>
       <name>fs.swift2d.service.SERVICE_NAME.tenant</name>
    	<value>TENANT</value>
	</property>
	<property>
       <name>fs.swift2d.service.SERVICE_NAME.password</name>
    	<value>PASSWORD</value>
	</property>
	<property>
       <name>fs.swift2d.service.SERVICE_NAME.username</name>
    	<value>USERNAME</value>
	</property>
	<property>
       <name>fs.swift2d.service.SERVICE_NAME.auth.method</name>
       <!-- swiftauth if needed -->
    	<value>keystone</value>
	</property>
	
	
#### Keystone V3 mapping to keys

| Driver configuration key | Keystone V3 key |
| ------------------------ | --------------- |
| fs.swift2d.service.SERVICE_NAME.username | user id |
| fs.swift2d.service.SERVICE_NAME.tenant | project id |

	<property>
	    <name>fs.swift2d.service.SERVICE_NAME.auth.url</name>
	    <value>https://identity.open.softlayer.com/v3/auth/tokens</value>
	</property>
	<property>
	    <name>fs.swift2d.service.SERVICE_NAME.public</name>
	    <value>true</value>
	</property>
	<property>
	    <name>fs.swift2d.service.SERVICE_NAME.tenant</name>
	    <value>PROJECTID</value>
	</property>
	<property>
	    <name>fs.swift2d.service.SERVICE_NAME.password</name>
	    <value>PASSWORD</value>
	</property>
	<property>
	    <name>fs.swift2d.service.SERVICE_NAME.username</name>
	    <value>USERID</value>
	</property>
	<property>
	    <name>fs.swift2d.service.SERVICE_NAME.auth.method</name>
	    <value>keystoneV3</value>
	</property>
	<property>
	    <name>fs.swift2d.service.SERVICE_NAME.region</name>
	    <value>dallas</value>
	</property>

###  Swift connector optional configuration

Below is the optional configuration that can be provided to Stocator and used internally to configure HttpClient.

| Configuration key | Default | Info |
| ------------------------ | --------------- |------|
| fs.stocator.MaxPerRoute | 25 | maximal connections per IP route |
| fs.stocator.MaxTotal | 50 | maximal concurrent connections |
| fs.stocator.SoTimeout | 1000 | low level socket timeout in milliseconds |
| fs.stocator.executionCount | 100 | number of retries for certain HTTP issues |
| fs.stocator.ReqConnectTimeout | 5000 | Request level connect timeout. Determines the timeout in milliseconds until a connection is established
| fs.stocator.ReqConnectionRequestTimeout | 5000 | Request level connection timeout. Returns the timeout in milliseconds used when requesting a connection from the connection manager
| fs.stocator.ReqSocketTimeout | 5000 | Defines the socket timeout (SO_TIMEOUT) in milliseconds, which is the timeout for waiting for data or, put differently, a maximum period inactivity between two consecutive data packets).
| fs.stocator.joss.synchronize.time | false | Will disable JOSS to synchronize time with the server. Setting this value to 'false' will badly impact on temp url. However this will reduce HEAD on account which might be problematic if the user doesn't has access rights to HEAD an account |
|fs.stocator.tls.version | false | if not provided, then system choosen is the default. In certain cases user may setup custom value, like TLSv1.2

## Configure Stocator's schemas (optional)
By default Stocator exposes `swift2d://` for Swift API and `cos://` for IBM Cloud Object Storage. It possible to configure Stocator to expose different schema.

The following example, will configure Stocator to respond to `swift://` in addition to `swift2d://`. This is useful, so users doesn't need to modify existing jobs that already uses hadoop-openstack connector. Below the example, how to configure Stocator to respond both to `swift://` and `swift2d://`

	<property>
		<name>fs.stocator.scheme.list</name>
		<value>swift2d,swift</value>
	</property>
	<!-- configure stocator as swift2d:// -->
	<property>
		<name>fs.swift2d.impl</name>
		<value>com.ibm.stocator.fs.ObjectStoreFileSystem</value>
	</property>
	<property>
		<name>fs.stocator.swift2d.impl</name>
		<value>com.ibm.stocator.fs.swift.SwiftAPIClient</value>
	</property>

release-notes.md
## Release notes

### Version 1.0.24

1. [cos] Fixing Spark streaming write flows, by adding support for object rename
2. [bug] Fixing globber. Filter accept method may skip slash of the parent, thus reject certain objects.

### Version 1.0.23

1. [bug] fixing Spark streaming write into COS
2. [swift] code cleanup
3. update http libraries

### Version 1.0.22

1. [cos] data cleanup for failed parts
2. [cos] adding tests for data cleanup
3. [cos] more unitests
4. [cos-ibm-sdk] support for token provided via url

### Version 1.0.21

1. [cos] improvements to fault tolerance algorithm
2. fixes to glober with brackets
3. [cos] array ouput stream. fixing package names

### Version 1.0.20

1. support for bracket globber mode
2. [test] adding COS tests for globber
3. [test] refactoring
4. Upgrade JOSS version from 0.10.1 to 0.10.2
5. [cos] list with UTF-8 to resolve + sign issues
6. remove getFileStatus call within mkdirs
7. update powermock dependency

### Version 1.0.19

1. [bug] Additional globber fixes
2. fixes to glober to bypass more patterns

### Version 1.0.18

1. 
[bug] fixing csv partitions that may fail with CSV files

### Version 1.0.17

1. [swift] Allow to configure TLS protocol version
2. [cos] ignore exception if stocator.properties not found
3. [cos] Fixing issues with unified object name extraction
4. Fixes to globber

### Version 1.0.16

1. Preserve file extensions for data parts
2. [swift] Allow Swift container name to contain spaces
3. Add shade generated pom to gitignore
4. [swift] Pass back failed keystone authentication failures
5. Update thirdparty dependencies to the latest compatible versions
6. [swift] Update PasswordScopeAccessProvider.java

### Version 1.0.15

1. [cos] Fixing BUFFER_DIR to provide input for the temp files folders
2. [swift, cos] Introducing object store flat globber
3. [cos] partial to support + in object names
4. avoid mkdirs on the directory that already exists. This resolved bug with TeraGen where mkdir writes into folder with existing data

### Version 1.0.14

1. [swift] Improvements for the connections (when data is ready and not on create)
2. Updates of the dependencies
3. [cos] Improve exceptions when IAM credentials are not valid
4. [cos] Fixing list on the root level
5. [cos] Usage of Statistics class to report metrics of bytes read or write
6. [cos] Remove preconditions check to avoid issues with dependencies

### Version 1.0.13

1. Using sha-256 for temp files. This prevents issues with long names

### Version 1.0.12

1. Fixing Null Pointer Exception when runnign with output comitter version 1
2. [cos] New configuration key to define Guava cache size
3. [cos] Fixing content type for block uploads

### Version 1.0.11
1. Improve object read flows
2. Allign list with Hadoop connectors. New configuration flag defines previous flat listing or new nested listing
3. Introduce Guava caching for frequent objects. This greatly reduces number of HEAD calls
4. Bucket names now can include dots
5. Support for partitions, introduced in Spark 2.0.X
6. Imrove temp file generation for write flows
7. Improve Object Store Globber
8. Improve response errors. This fixed bug that reported NPE instead of not authenticated.
9. Block upload for COS connector. Based on disk
10. Implementation for isFile / isDirectory methods

### Version 1.0.10
1. Move to Hadoop 2.7.3
2. Code cleanup
3. Remove dependance on FSExceptionMessages
4. Adjust copyright headers
5. Fix parallel bucket creation
6. Custom user agent

###Version 1.0.11
1. Improve object read flows
2. Allign list with Hadoop connectors. New configuration flag defines previous flat listing or new nested listing
3. Introduce Guava caching for frequent objects. This greatly reduces number of HEAD calls
4. Bucket names now can include dots
5. Support for partitions, introduced in Spark 2.0.X
6. Imrove temp file generation for write flows
7. Improve Object Store Globber
8. Improve response errors. This fixed bug that reported NPE instead of not authenticated.
9. Block upload for COS connector. Based on disk
10. Implementation for isFile / isDirectory methods


### Version 1.0.10
1. Move to Hadoop 2.7.3
2. Code cleanup
3. Remove dependance on FSExceptionMessages
4. Adjust copyright headers
5. Fix parallel bucket creation
6. Custom user agent

### Version 1.0.9
1. Stocator COS support 
2. Fixing Stocator user agent
3. Fixing issues with streaming
4. Proper handle of the Filter for list operations
4. Moving JOSS to 0.9.15
5. Avoid duplicate get container
6. Making API to work with non US locale

### Version 1.0.8
1. Better debug prints
2. Reducing number of GET requests
3. Fixing list status
4. Fixing get file status on temp object
5. Remove duplicate call to get object length
6. Support for temp urls
7. Added thread pool for create method
8. Support spaces in the names

### Version 1.0.7
1. Modified JOSS to disable HEAD on account when accessing containers. This caused issues when user doesn't has access on account level, but only on container level.
2. Fixed regression caused by consumeQueitely. This fix improved read by 3 times
3. Added cache to contain object length and last modified time stamp. This cache is filled during list and usefull for Spark flows.
4. Removed need to HEAD object before GET. This reduces number of HEAD requests.

### Version 1.0.6
1. Continue improvements with container listing
2. Object upload now based on the Apache HttpClient 4.5.2
3. New configuration keys to tune connection properties
4. Moving Hadoop to 2.7.2
5. Adapting Stocator to work with Hadoop testDFSIO. This includes support for certain flows that required by Hadoop.
6. Continue improvements to logging.

### Version 1.0.5

1. Fixing object store globber. Resolving issues with  container listings
2. Introducing SwiftConnectionManager that is based on  PoolingHttpClientConnectionManager. This makes better connection utilizations both for SwiftAPIDirect and JOSS.

src/main/java/com/ibm/stocator/fs/cos/COSConstants.java
/**
 * (C) Copyright IBM Corp. 2015, 2016
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.ibm.stocator.fs.cos;

/**
 * Constants used in the COS REST protocol.
 */
public class COSConstants {
  /*
   * COS name space identifier
   */
  public static final String S3_D = "s3d";
  public static final String S3_A = "s3a";
  public static final String COS = "cos";

  /*
   * COS configuration prefix in the core-site.xml
   */
  public static final String FS_S3_D = "fs." + S3_D;
  public static final String FS_COS = "fs." + COS;
  public static final String FS_S3_A = "fs." + S3_A;
  /*
   * COS configuration key in the core-site.xml
   */
  public static final String S3_D_SERVICE_PREFIX = FS_S3_D + ".";
  public static final String S3_A_SERVICE_PREFIX = FS_S3_A + ".";
  public static final String COS_SERVICE_PREFIX = FS_COS + ".";

  public static final String COS_BUCKET_PROPERTY = FS_S3_D + ".BUCKET-NAME";

  public static final String REGION = ".region.name";
  public static final String REGION_COS_PROPERTY = FS_COS + REGION;

  public static final String ACCESS_KEY = ".access.key";
  public static final String ACCESS_KEY_COS_PROPERTY = FS_COS + ACCESS_KEY;

  public static final String API_KEY = ".iam.api.key";
  public static final String API_KEY_IAM_PROPERTY = FS_COS + API_KEY;

  public static final String IAM_TOKEN = ".iam.token";
  public static final String IAM_TOKEN_PROPERTY = FS_COS + IAM_TOKEN;

  public static final String IAM_ENDPOINT = ".iam.endpoint";
  public static final String IAM_ENDPOINT_PROPERTY = FS_COS + IAM_ENDPOINT;

  public static final String IAM_SERVICE_INSTANCE_ID = ".iam.service.id";
  public static final String IAM_SERVICE_INSTANCE_ID_PROPERTY = FS_COS + IAM_SERVICE_INSTANCE_ID;

  public static final String SECRET_KEY = ".secret.key";
  public static final String SECRET_KEY_COS_PROPERTY = FS_COS + SECRET_KEY;

  public static final String ENDPOINT_URL = ".endpoint";
  public static final String ENDPOINT_URL_COS_PROPERTY = FS_COS + ENDPOINT_URL;

  public static final String BLOCK_SIZE = ".block.size";
  public static final String BLOCK_SIZE_COS_PROPERTY = FS_COS + BLOCK_SIZE;

  public static final String AUTO_BUCKET_CREATE = ".create.bucket";
  public static final String AUTO_BUCKET_CREATE_COS_PROPERTY = FS_COS + AUTO_BUCKET_CREATE;

  public static final String V2_SIGNER_TYPE = ".v2.signer.type";
  public static final String V2_SIGNER_TYPE_COS_PROPERTY = FS_COS + V2_SIGNER_TYPE;

  public static final String INPUT_POLICY = ".input.policy";
  public static final String INPUT_POLICY_COS_PROPERTY = FS_COS + INPUT_POLICY;

  public static final String READAHEAD_RANGE = ".readahead.range";
  public static final long DEFAULT_READAHEAD_RANGE = 64 * 1024;

  public static final String SOCKET_TIMEOUT = ".connection.timeout";
  public static final int DEFAULT_SOCKET_TIMEOUT = 800000;

  public static final String SOCKET_SEND_BUFFER = ".socket.send.buffer";
  public static final int DEFAULT_SOCKET_SEND_BUFFER = 8 * 1024;

  public static final String SOCKET_RECV_BUFFER = ".socket.recv.buffer";
  public static final int DEFAULT_SOCKET_RECV_BUFFER = 8 * 1024;

  public static final String MAX_PAGING_KEYS = ".paging.maximum";
  public static final int DEFAULT_MAX_PAGING_KEYS = 5000;

  // the maximum number of threads to allow in the pool used by TransferManager
  public static final String MAX_THREADS = ".threads.max";
  public static final int DEFAULT_MAX_THREADS = 10;

  // the time an idle thread waits before terminating
  public static final String KEEPALIVE_TIME = ".threads.keepalivetime";
  public static final int DEFAULT_KEEPALIVE_TIME = 60;

  //override signature algorithm used for signing requests
  public static final String SIGNING_ALGORITHM = ".signing-algorithm";

  // number of simultaneous connections to cos
  public static final String MAXIMUM_CONNECTIONS = ".connection.maximum";
  public static final int DEFAULT_MAXIMUM_CONNECTIONS = 10000;

  //number of times we should retry errors
  public static final String MAX_ERROR_RETRIES = ".attempts.maximum";
  public static final int DEFAULT_MAX_ERROR_RETRIES = 20;

  //seconds until we give up trying to establish a connection to cos
  public static final String ESTABLISH_TIMEOUT = ".connection.establish.timeout";
  public static final int DEFAULT_ESTABLISH_TIMEOUT = 50000;

  // size of each of or multipart pieces in bytes
  public static final String MULTIPART_SIZE = ".multipart.size";
  public static final long DEFAULT_MULTIPART_SIZE = 8388608; // 8 MB

  // minimum size in bytes before we start a multipart uploads or copy
  public static final String MIN_MULTIPART_THRESHOLD = ".multipart.threshold";
  public static final long DEFAULT_MIN_MULTIPART_THRESHOLD = Integer.MAX_VALUE;

  // the maximum number of tasks cached if all threads are already uploading
  public static final String MAX_TOTAL_TASKS = ".max.total.tasks";
  public static final int DEFAULT_MAX_TOTAL_TASKS = 5;

  //comma separated list of directories
  public static final String BUFFER_DIR = ".buffer.dir";

  // amount of time (in ms) to allow the client to complete the execution of
  // an API call
  public static final String CLIENT_EXEC_TIMEOUT = ".client.execution.timeout";
  public static final int DEFAULT_CLIENT_EXEC_TIMEOUT = 500000;

  // amount of time to wait (in ms) for the request to complete before giving
  // up and timing out
  public static final String REQUEST_TIMEOUT = ".request.timeout";
  public static final int DEFAULT_REQUEST_TIMEOUT = 500000;
  // connect to cos over ssl?
  public static final String SECURE_CONNECTIONS = ".connection.ssl.enabled";
  public static final boolean DEFAULT_SECURE_CONNECTIONS = true;
  //connect to cos through a proxy server?
  public static final String PROXY_HOST = ".proxy.host";
  public static final String PROXY_PORT = ".proxy.port";
  // of the form proxy_ip:proxy_port
  public static final String PROXY_ENV_HTTP_PROXY = ".env.http.proxy";
  public static final String PROXY_ENV_HTTPS_PROXY = ".env.https.proxy";
  public static final String PROXY_USERNAME = ".proxy.username";
  public static final String PROXY_PASSWORD = ".proxy.password";
  public static final String PROXY_DOMAIN = ".proxy.domain";
  public static final String PROXY_WORKSTATION = ".proxy.workstation";

  // User agent prefix
  public static final String USER_AGENT_PREFIX = ".user.agent.prefix";
  public static final String DEFAULT_USER_AGENT_PREFIX = "";

  // community block upload support
  public static final String FAST_UPLOAD_BUFFER_ARRAY = "array";
  public static final String FAST_UPLOAD_BUFFER_DISK = "disk";
  public static final String FAST_UPLOAD_BYTEBUFFER = "bytebuffer";

  public static final String FAST_UPLOAD = ".fast.upload";
  public static final boolean DEFAULT_FAST_UPLOAD = false;

  public static final String FAST_UPLOAD_BUFFER =
      ".fast.upload.buffer";
  public static final String DEFAULT_FAST_UPLOAD_BUFFER =
      FAST_UPLOAD_BUFFER_DISK;

  public static final String FAST_UPLOAD_ACTIVE_BLOCKS =
      ".fast.upload.active.blocks";
  public static final int DEFAULT_FAST_UPLOAD_ACTIVE_BLOCKS = 4;

  /** The minimum multipart size which COS supports. */
  public static final int MULTIPART_MIN_SIZE = 5 * 1024 * 1024;
  public static final int MAX_MULTIPART_COUNT = 10000;
  //enable multiobject-delete calls?
  public static final String ENABLE_MULTI_DELETE =
      ".multiobjectdelete.enable";

  // should we try to purge old multipart uploads when starting up
  public static final String PURGE_EXISTING_MULTIPART =
      ".multipart.purge";
  public static final boolean DEFAULT_PURGE_EXISTING_MULTIPART = false;

  // purge any multipart uploads older than this number of seconds
  public static final String PURGE_EXISTING_MULTIPART_AGE =
      ".multipart.purge.age";
  public static final long DEFAULT_PURGE_EXISTING_MULTIPART_AGE = 86400;

  public static final String FLAT_LISTING = ".flat.list";
  public static final boolean DEFAULT_FLAT_LISTING = true;

  public static final String INPUT_FADVISE = "experimental.input.fadvise";
  public static final String INPUT_FADV_NORMAL = "normal";
  public static final String INPUT_FADV_SEQUENTIAL = "sequential";
  public static final String INPUT_FADV_RANDOM = "random";

  public static final String IAM_TOKEN_MAX_RETRY = ".iam.token.max.retry";
  public static final String IAM_TOKEN_MAX_RETRY_PROPERTY = FS_COS + IAM_TOKEN_MAX_RETRY;

  public static final String IAM_TOKEN_REFRESH_OFFSET = ".iam.token.refresh";
  public static final String IAM_TOKEN_REFRESH_OFFSET_PROPERTY = FS_COS + IAM_TOKEN_REFRESH_OFFSET;

}
